website/blog/2024-10-25-llama-index-workflow/index.md

```diff
@@ -0,0 +1,423 @@
+
+```python
+    reranker = RankGPTRerank(llm=self.llm, top_n=5)
+    reranked_nodes = reranker.postprocess_nodes(ev.nodes, query_str=query)
+    reranked_context = "\n".join(node.text for node in reranked_nodes)
+```
+
+Finally, the reranked context is passed to the LLM along with the user query to generate the final answer. The result is returned as a `StopEvent` with the `result` key.
+
+```python
+    @step
+    async def query_result(self, ctx: Context, ev: QueryEvent) -> StopEvent:
+        """Get result with relevant text."""
+        query = await ctx.get("query")
+
+        prompt = FINAL_QUERY_TEMPLATE.format(context=ev.context, query=query)
+        response = self.llm.complete(prompt).text
+        return StopEvent(result=response)
+```
+
+Now, let's instantiate the workflow and run it.
+
+```python
+# Workflow with VS + BM25 retrieval
+from workflow.workflow import HybridRAGWorkflow
+
+workflow = HybridRAGWorkflow(retrievers=["vector_search", "bm25"], timeout=60)
+response = await workflow.run(query="Why use MLflow with LlamaIndex?")
+print(response)
+```
+
+## 6. Log the Workflow in MLflow Experiment
+
+Now we want to run the workflow with various different retrieval strategy and evaluate the performance. However, before running the evaluation, we'll log the model in MLflow to track both the model and its performance within an **MLflow Experiment**.
+
+For the LlamaIndex Workflow, we use the new [Model-from-code](https://mlflow.org/docs/latest/models.html#models-from-code) method, which logs models as standalone Python scripts. This approach avoids the risks and instability associated with serialization methods like pickle, relying instead on code as the single source of truth for the model definition. When combined with MLflow's environment-freezing capability, it provides a reliable way to persist models. For more details, refer to the [MLflow documentation](https://mlflow.org/docs/latest/models.html#models-from-code).
+
+💡 In the `workflow` directory, there's a `model.py` script that imports the `HybridRAGWorkflow` and instantiates it with dynamic configurations passed via the `model_config` parameter during logging. This design allows you to track models with different configurations without duplicating the model definition.
+
+We'll start an MLflow Run and log the model script `model.py` with different configurations using the [mlflow.llama_index.log_model()](https://mlflow.org/docs/latest/python_api/mlflow.llama_index.html#mlflow.llama_index.log_model) API.
+
+
+```python
+# Different configurations we will evaluate.
+run_name_to_retrievers = {
```

Comment:
How much overkill is it to do all valid combinations?

```
run_name_to_retrievers = {
    # 1. No retrievers (prior knowledge in LLM).
    "none": [],
    # 2. Vector search retrieval.
    "vs": ['vector_search'],
    # 3. Keyword search (BM25).
    "bm25": ['bm25'],
    # 4. Web search.
    "web": ['web_search'],
    # 5. Vector search retrieval and Keyword search (BM25).
    "vs + bm25": ['vector_search', 'bm25'],
    # 6. Vector search retrieval and Web search.
    "vs + web": ['vector_search', 'web_search'],
    # 7. Keyword search (BM25) and Web search.
    "bm25 + web": ['bm25', 'web_search'],
    # 8. Vector search retrieval and Keyword search (BM25) and Web search.
    "vs + bm25 + web": ['vector_search', 'bm25', 'web_search'],
}
```

Reply:
Good point, I initially did it but each combination costed some $$ and minutes so I shrank it😅 
Let me add a comment that we only have 4 here for demonstration purpose but users are not limited to run more.

Reply:
Comment sounds perfect :D